ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/ConnectionFactory.kt

```diff
@@ -22,14 +22,18 @@ internal class ConnectionFactory(
         configuration: SocketOptions.TCPClientSocketOptions.() -> Unit = {}
     ): Socket {
         limit.acquire()
-        val addressSemaphore = addressLimit.computeIfAbsent(address) { Semaphore(addressConnectionsLimit) }
-        addressSemaphore.acquire()
-
         return try {
-            aSocket(selector).tcpNoDelay().tcp().connect(address, configuration)
+            val addressSemaphore = addressLimit.computeIfAbsent(address) { Semaphore(addressConnectionsLimit) }
+            addressSemaphore.acquire()
+
+            try {
```

Comment:
Why nested `try`/`catch` block is necessary here?

Reply:
Hi @rsinukov, thanks for the review.

> Why nested try/catch block is necessary here?

It's to make sure that the `limit` semaphore is released when failure.

In the original code, there's no try-catch between the `limit` acquisition and `addressSemaphore` acquisition, so if it succeeds to acquire `limit` semaphore but it fails to acquire the `addressSemaphore` (e.g. by timeout), then it didn't release the `limit` semaphore - this is the resource leak.

> Can you explain steps in which this bug can be reproduced?

With the original code, this new test I added will be blocked forever.

https://github.com/ktorio/ktor/blob/7c08c9cb359d8196f6adb218cf6d1a8282e00fa5/ktor-client/ktor-client-cio/jvmAndNix/test/ConnectionFactoryTest.kt#L60

Because [this failure](https://github.com/ktorio/ktor/blob/7c08c9cb359d8196f6adb218cf6d1a8282e00fa5/ktor-client/ktor-client-cio/jvmAndNix/test/ConnectionFactoryTest.kt#L72) doesn't release the `limit` semaphore, this attempt cannot acquire the `limit` semaphore.

Reply:
Thanks for the explanation. But isn't it enough just to move `addressSemaphore.acquire()` into the outer `try/catch` block?

Reply:
Do you mean, something like this?

```kotlin
limit.acquire()
val addressSemaphore = try {
  addressLimit.computeIfAbsent(address) { Semaphore(addressConnectionsLimit) }
} catch (cause: Throwable) {
  limit.release()
  throw cause
}

try {
  aSocket(selector).tcpNoDelay().tcp().connect(address, configuration)
} catch (cause: Throwable) {
  // a failure or cancellation
  addressSemaphore.release()

  limit.release() // <= we need limit.release() here as well

  throw cause
}
```

We can, but I believe nested try-catch is more straightforward. Once `limit` is acquired, we need to release it whenever error happens in this method.

Reply:
No, i mean just moving `addressSemaphore.acquire()` inside the `try/catch` and leave everything else as is. Since it's the only suspendible call, only it can throw an exception. And there is no point in having one extra nested block
```
        limit.acquire()
        val addressSemaphore = addressLimit.computeIfAbsent(address) { Semaphore(addressConnectionsLimit) }
        return try {
            addressSemaphore.acquire()
            aSocket(selector).tcpNoDelay().tcp().connect(address, configuration)
        } catch (cause: Throwable) {
            limit.release()
            throw cause
        }
```

Reply:
Hmm, it looks it will release `addressSemaphore` when it fails to `addressSemaphore.acquire` (by cancellation while waiting for the semaphore acquisition for example) - which means `addressSemaphore` will accidentally release more semaphore than it should 🤔